### ISSUE TITLE
Multiple un-necessary calls to FlareSolverr instead of re-using the cookies? (e.g. form relogin for CardigannIndexer)

### ISSUE THREAD

**Issue by edvincent** [Jackett/Jackett#10824]
TL;DR: Jackett triggering three requests to FlareSolverr, causing delays, and skyrocketing the possibilities that we hit a captcha.

### Environment

All of them as this is the code that is written to do this. ;D

### Description

Was looking at the logs of FlareSolverr, and realized that for each "Test" I run from the UI for a basic indexer with a form (e.g. YGGTorrent), three unique Cloudflare challenges are detected in FlareSolverr. This triplicates the possibilities of getting a Captcha, and I'm pretty sure 3 completely independent calls within 15 secs from the same IP is the kind of stuff that Cloudlfare uses to flag "automated" behavior, which itself. There's also the having to wait 3 times more, which is a pain when doing manual searches for examples.

As far as I am concerned, it's extremely rare that ONE of those three calls doesn't get a Captcha, which effectively means nothing works; and this indexer has a very small cookie time so you better be lucky... Coupled with the exponential backoff from Radarr/Sonarr, this indexer is virtually always unavailable for me.

The three calls that trigger FlareSolverr that I've been able to trace down would be:
1. The initial query, that triggers what I consider should be the only clearance request from FlareSolverr.
2. The call to get a Configuration for setup and an empty page for the log form, which [purposely drops any previous cookies](https://github.com/edvincent/Jackett/blob/master/src/Jackett.Common/Indexers/CardigannIndexer.cs#L840), ending up with a new request to FlareSolverr.
3. The third one seems to be the call to the new query. Honestly, I'm a bit clueless why if we never delete the cookies (i.e. we merge and keep cookies from (1) and (2)), this doesn't make a new request to FlareSolverr. The only difference that I see is that it collects a `ga_` cookie on top of an already-existing `ga` cookie.

### Some logs showing this

Simply added some more logging for me to understand the flow... `VALUE_X` are from the first request (which we lose completely) and `VALUE_1X` are from the second request. Request 1, 2 and 4 are the ones triggering FlareSolverr.

```
01-15 01:39:11 Info request to https://www2.yggtorrent.si/engine/search?category=all&name=&description=&file=&uploader=&sub_category=&do=search&order=desc&sort=publish_date with Cookies
01-15 01:39:18 Info updating Cookies  => ygg_=VALUE_1; cf_chl_1=VALUE_2; cf_clearance=VALUE_3; __cfduid=VALUE_4; cf_chl_prog=VALUE_5; ga=VALUE_6
01-15 01:39:19 Info CardigannIndexer (yggtorrent): Relogin required
01-15 01:39:19 Info request to https://www.yggtorrent.si/ with Cookies
01-15 01:39:25 Info request to https://www.yggtorrent.si/ with Cookies ygg_=VALUE_10; ga_=VALUE_11; cf_clearance=VALUE_12; __cfduid=VALUE_13
01-15 01:39:26 Info request to https://www2.yggtorrent.si/engine/search?category=all&name=&description=&file=&uploader=&sub_category=&do=search&order=desc&sort=publish_date with Cookies ygg_=VALUE_10; ga_=VALUE_11; cf_clearance=VALUE_12; __cfduid=VALUE_13
01-15 01:39:34 Info request to https://www2.yggtorrent.si/engine/search?category=all&name=&page=50&description=&file=&uploader=&sub_category=&do=search&order=desc&sort=publish_date with Cookies ygg_=VALUE_10; ga_=VALUE_11; cf_clearance=VALUE_12; __cfduid=VALUE_13; ga=VALUE_14
```

### What next?

I'm happy to try and take a stab at it, but I need someone to help me understand some history that I might not see/know myself.

1. Is there any reason why we need to drop any previous cookies in [GetConfigurationForSetup](https://github.com/edvincent/Jackett/blob/master/src/Jackett.Common/Indexers/CardigannIndexer.cs#L840)? I was thinking maybe it'd cause issues if we presented a past, expired cookie for the indexer (preventing us from logging) but I tested locally and it works flawlessly... 
2. If we need to drop some cookies... Would you be happy with somehow keeping a list of Cloudflare cookies that we don't want to drop?
3. The other idea I was toying with was looking at FlareSolverr's session concept, but I don't think FlareSolverrSharp has it integrated yet... The idea being to keep the session open until the end of the query, in case we need to login again...

Open to any ideas/suggestions otherwise.

**Comment by edvincent** [comment]
Aaaaaand I'll immediately reply to myself for the request (3). This seems to be an issue with that `ga` vs `ga_` cookie, which is itself due to [different subdomains being used in the default config](https://github.com/Jackett/Jackett/blob/master/src/Jackett.Common/Definitions/yggtorrent.yml). I guess it somehow doesn't like the switch from `www` to `www2`, and needs a new Google Analytics cookies which somehow is messing with Cloudflare as it makes it present a new challenge?

Using `www2` in both configuration entries for SiteLink and Search and Download URL gets rid of that 3rd request to FlareSolverr and no `ga_` cookie to be generated.

**Comment by ngosang** [comment]
> Is there any reason why we need to drop any previous cookies in GetConfigurationForSetup? I was thinking maybe it'd cause issues if we presented a past, expired cookie for the indexer (preventing us from logging) but I tested locally and it works flawlessly...

Jackett support more than 500 indexers and they share most of the code. I guess keeping the old cookies can cause problems in other trackers but I'm not sure.

> The other idea I was toying with was looking at FlareSolverr's session concept, but I don't think FlareSolverrSharp has it integrated yet... The idea being to keep the session open until the end of the query, in case we need to login again...

I have been thinking in this but keeping Chrome open using CPU and 500MB - 1 GB of RAM just for this case is overkill.


@ilike2burnthing is there any reason to have 2 URLs in YGGtorrent indexer? If we use only 1 we can reduce the number of requests to FlareSolverr. 
I don't use this tracker but if the 3 request are only in configuration the issue is not really big. How many requests it makes in one regular day (without touching the configuration?
`docker logs flaresolverr | grep Param`

**Comment by edvincent** [comment]
> Jackett support more than 500 indexers and they share most of the code. I guess keeping the old cookies can cause problems in other trackers but I'm not sure.

Yeah I was thinking the same about other trackers, but at the same time, that's exactly what a browser would do... So if keeping cookies issued 5s ago was bad here, it'd also be causing issues when trying to browse the website... I think the only difference would be that a browser wouldn't send an expired cookie, but here we're literally talking seconds... I could build something that drops the expired cookies and keep the rest; but doubt anything sends couple-of-seconds-worth cookies?

The other ways is to start tracking cookies... We know that keeping only the Cloudflare cookies is not bad (and helps, and what I'm trying to achieve :p), so we could simply hard-code the CloudFlare cookie name(s) and drop everything but those... 

> I have been thinking in this but keeping Chrome open using CPU and 500MB - 1 GB of RAM just for this case is overkill.

Well not really. I don't mean keeping it open forever, but for those 10-15s that several calls need to be made. Something like [creating a session when making the query call](https://github.com/Jackett/Jackett/blob/master/src/Jackett.Common/Indexers/CardigannIndexer.cs#L1269), then [passing its ID if we need to DoLogin() again](https://github.com/Jackett/Jackett/blob/master/src/Jackett.Common/Indexers/CardigannIndexer.cs#L1290), and always closing the session [before returning the releases](https://github.com/Jackett/Jackett/blob/master/src/Jackett.Common/Indexers/CardigannIndexer.cs#L1625). (accounting for the exceptions that might prematurely stop the query, obviously).

However, that'd effectively be the same as keeping the cookies as I was suggesting above, and keeping cookies would be waaaaay easier to implement.

> @ilike2burnthing is there any reason to have 2 URLs in YGGtorrent indexer? If we use only 1 we can reduce the number of requests to FlareSolverr.

I'm already double-checking with people in that tracker, but I've changed it on my Jackett config and no issues reaching out to the same URL. I'll send a PR for the 2 URLs in YGGtorrent if confirmed to cause no issues.

**Comment by ilike2burnthing** [comment]
www.* is used for login, www2.* doesn't work with the current indexer. You can see the whole mess leading up to it being implemented here - https://github.com/Jackett/Jackett/issues/4504

Previously the post method with `/user/login` was used (which may work for www2.*), but removed due to problems - https://github.com/Jackett/Jackett/issues/2852. I'll give it a try to see if anything has changed since then.

Under normal functioning, there's only one request made to www2.*

**Comment by edvincent** [comment]
FWIW, I've had ```www2``` running in my Jackett for couple of days with no issues, and only 2 requests on re-login.

Even trying manually on the browser, when you logout (after a login from any domain, like even ```www```) you get forced to ```www2```. If you try to login from there, the POST request to ```/user/login``` goes to ```www2``` and you get a 200 no probs. I'll try to read-up those two past issues, but that's exactly what I needed to get, some context =D

#### Login request from Chrome's Network Console

```
Request URL: https://www2.yggtorrent.si/user/login
Request Method: POST
Status Code: 200 
```

```
:authority: www2.yggtorrent.si
:method: POST
:path: /user/login
:scheme: https
origin: https://www2.yggtorrent.si
referer: https://www2.yggtorrent.si/
```

**Comment by ilike2burnthing** [comment]
I'll post a PR in a few mins, all comments welcome

**Comment by ilike2burnthing** [comment]
https://github.com/Jackett/Jackett/pull/10852

**Comment by xfouloux** [comment]
i've put your modified yggtorrent.yml on my production jackett, to see if this is causing less issue 

I have to recheck the indexer sometime 3 or 4 times a day because sonarr/radarr puts it in not working state, let's hope it fixes that at least a bit.

**Comment by xfouloux** [comment]
well it does not break anything, seems to work fine !
I still have to recheck the indexer manually at least 1 time a day, but that is because of flaresolverr not being able to complete the captcha sometimes, can't do anything about that, but like it was said, like this you have less chances to get the captcha i guess

**Comment by edvincent** [comment]
Yep same here, no probs overnight with the YAML from the PR, seeing less overall failures which is a great first step! 3 => 2 calls.

For the 2 => 1 calls we were discussing cause by dropping headers, if we want to be on the safe-side (i.e. keep the same behavior for everything existing), I'm proposing a YAML option that CardigannIndexer parses which is a simple true/false (defaults to false) to keep all headers when logging-in?

@ilike2burnthing @ngosang (and others) If the above approach sounds good to you (or others from [1] sound better), I'll send a PR with that in the next couple of days.

FWIW, I'll then likely start looking at some kind of retries (with a random delay to avoid being flagged as a bot, within the timeout) when we get a captcha; as locally I've found 1 or 2 calls after I do get a clearance instead of a captcha. That's also what people effectively achieve by testing manually several times until it works and they can search for their stuff...

------

[1] The other ideas I had but not convinced:
1. Hard-code the CloudFlare cookie names (```cf_chl_1```, ```cf_clearance```, ```__cfduid```, ```cf_chl_prog```, and any new one introduced) and keep them when logging-in. => _I like it better than the true/false from above (as effectively we know it's safe to keep those) and will benefit every indexer; but I also despise hard-coding stuff that will break everything when it changes._
1. Introduce a YAML option for CardigannIndexer that will contain a comma-separated list of names of cookies the code should **drop** when logging-in. => _Not great because it's not backwards compatible. Indexers that could have some kind of issues from keeping cookies would be broken by the change until someone changed their YAML._
1. Introduce a YAML option for CardigannIndexer that will contain a comma-separated list of names of cookies the code should **not drop** when logging-in. => _Not great because effectively every indexer would need to duplicate the CloudFlare cookie names._

**Comment by ilike2burnthing** [comment]
I'll leave ngosang to comment, but from a layman's perspective your first suggestion of adding an option which defaults to false (i.e. everything would stay as it is for all current indexers), sounds fine.

The only indexers I'm aware of using delays and retries are written in C#, so I'm not sure if this can be replicated in YAML, but have at it!

**Comment by ngosang** [comment]
I'm happy to merge #10852 

> For the 2 => 1 calls we were discussing cause by dropping headers, if we want to be on the safe-side (i.e. keep the same behavior for everything existing), I'm proposing a YAML option that CardigannIndexer parses which is a simple true/false (defaults to false) to keep all headers when logging-in?

I prefer to no implement this for now. In most pages the user session (cookie after login ) is valid for several days. First we have to review why this is not the case in YGG.

**Comment by ilike2burnthing** [comment]
[v0.17.333](https://github.com/Jackett/Jackett/releases/tag/v0.17.333)

**Comment by ilike2burnthing** [comment]
Just a heads up Re: using form login again (though still only using the single site link) - https://github.com/Jackett/Jackett/commit/bf2d8f8ded99868a70b7321d9f4e62f29fb93ed7

The site has dropped Cloudflare DDoS Protection (for now), revealing an issue with the site redirecting from `/user/login` to `/user/login?attempt=1`, preventing Jackett from logging in (which I'm guessing Cloudflare was allowing us to work around).

Having dropped Cloudflare though, yggcookie is now working again, downloads and all.

**Comment by ngosang** [comment]
Fixed in 
FlareSolverr 2.2.6 => https://github.com/FlareSolverr/FlareSolverr/releases/tag/v2.2.6
Jackett v0.20.1496 => https://github.com/Jackett/Jackett/releases/tag/v0.20.1496
